### PR TITLE
Prohibit Option#toString and io.vavr.collection.<X>#peek using FenumChecker

### PR DESCRIPTION
**Pull request by PawelLipski**
~Depends on changes from https://github.com/antkorwin/better-strings/pull/14/files~ => changes dependent on better-strings v0.4 were moved to #362

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -132,6 +132,7 @@ subprojects {
     // See https://github.com/kelloggm/checkerframework-gradle-plugin/issues/85 (and the associated PR #86) for why this flag is needed
     suppressLombokWarnings = false
     checkers = [
+      'org.checkerframework.checker.fenum.FenumChecker',
```
Comment (PawelLipski) [inline build.gradle:135]: Possibly can be eventually moved to CI-only checkers, as with `IndexChecker` & others below

File: build.gradle [hunk 2]
```diff
@@ -147,6 +148,7 @@ subprojects {
       '-AassumeAssertionsAreEnabled',
       '-AinvariantArrays',
       '-Alint=cast:redundant,cast:unsafe',
+      '-ANullnessChecker_lint=redundantNullComparison',
```
Comment (PawelLipski) [inline build.gradle:151]: Unrel

File: build.gradle [hunk 3]
```diff
@@ -199,12 +201,22 @@ subprojects {
   // This needs to be enabled in each subproject by default because there's going to be no warning
   // if this annotation processor isn't run in any subproject (the strings will be just interpreted verbatim, without interpolation applied).
   // We'd only capture that in CI's post-compile checks by analyzing constants in class files.
+  repositories {
```
Comment (PawelLipski) [inline build.gradle:204]: To be removed if/when `better-strings` are published

File: build.gradle [hunk 4]
```diff
@@ -308,7 +320,7 @@ subprojects {
       checkerFramework {
         checkers += 'org.checkerframework.common.subtyping.SubtypingChecker'
         def qualClassDir = project(':qual').sourceSets.main.output.classesDirs.asPath
-        extraJavacArgs += "-AqualDirs=${qualClassDir}"
+        extraJavacArgs += "-ASubtypingChecker_qualDirs=${qualClassDir}"
```
Comment (PawelLipski) [inline build.gradle:323]: `-AXChecker_y` means "only apply `-Ay` to `XChecker`"

File: gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreRepository.java [hunk 5]
```diff
@@ -405,11 +405,10 @@ public boolean isAncestor(IGitCoreCommit presumedAncestor, IGitCoreCommit presum
...
       LOG.debug("Starting revwalk");
       return Iterator.ofAll(walk.iterator())
-          .takeUntil(revCommit -> revCommit.getId().getName().equals(untilExclusive.getHash().getHashString()))
-          .map(revCommit -> {
-            LOG.debug(() -> "* " + revCommit.getId().getName());
-            return new GitCoreCommit(revCommit);
-          })
+          .toJavaStream()
+          .takeWhile(revCommit -> !revCommit.getId().getName().equals(untilExclusive.getHash().getHashString()))
+          .peek(revCommit -> LOG.debug(() -> "* " + revCommit.getId().getName()))
```
Comment (PawelLipski) [inline gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreRepository.java:410]: `peek` on Java Stream, frowned upon but allowed

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BaseRebaseBranchOntoParentAction.java [hunk 6]
```diff
@@ -82,20 +82,20 @@ public void update(AnActionEvent anActionEvent) {
 
         if (anActionEvent.getPlace().equals(ActionPlaces.ACTION_PLACE_TOOLBAR)) {
           presentation.setEnabled(false);
-          presentation.setDescription("Root branch '${branchName}' cannot be rebased");
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BaseRebaseBranchOntoParentAction.java:85]: Whoops, it turns out we had a few bugs :beetle:

File: build.gradle [hunk 7]
```diff
@@ -200,12 +202,22 @@ subprojects {
...
+
   dependencies {
     def betterStringsArtifact = [ group: 'com.antkorwin', name: 'better-strings', version: betterStringsVerson ]
     annotationProcessor(betterStringsArtifact)
     testAnnotationProcessor(betterStringsArtifact)
   }
 
+  gradle.projectsEvaluated {
+    tasks.withType(JavaCompile) {
+      options.compilerArgs += '-AcallToStringExplicitlyInInterpolations'
```
Comment (PawelLipski) [inline build.gradle:217]: To make sure that the generated interpolated code generates `xd.toString()` and not `String.valueOf(xd)`

Comment (PawelLipski) [reply build.gradle:217]: Actually, adding this option is the subject of my PR to better-strings

File: config/checker/java.util.astub [hunk 8]
```diff
@@ -8,6 +8,10 @@ import org.checkerframework.dataflow.qual.Pure;
 
 package java.util;
 
+class Objects {
+  static <T> @NonNull T requireNonNull(@Nullable T obj);
```
Comment (PawelLipski) [inline config/checker/java.util.astub:12]: Required to make sure that the code generated for interpolations in`-AcallToStringExplicitlyInInterpolations` mode (which uses `requireNonNull` under the hood) doesn't raise warnings with Checker

File: scripts/enforce-pre-release-version-correct-use [hunk 9]
```diff
@@ -28,6 +28,6 @@ elif [[ ${CIRCLE_BRANCH-} ]]; then
       die 'Pre-release version is not allowed on master/release/hotfix/backport branch tip'
   else
     [[ $head_pre_release ]] || \
-      die 'Non-master/release/hotfix/backport branch tip must have a pre-release version (a.b.c-d, not a.b.c)'
+      die 'Non-{master/release/hotfix/backport} branch tip must have a pre-release version (a.b.c-d, not a.b.c)'
```
Comment (PawelLipski) [inline scripts/enforce-pre-release-version-correct-use:31]: Fixed as requested on #294 ;)

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRebaseBranchOntoParentAction.java [hunk 10]
```diff
@@ -70,28 +70,24 @@ public void update(AnActionEvent anActionEvent) {
...
 
       } else if (branch.get().asNonRootBranch().getSyncToParentStatus().equals(SyncToParentStatus.MergedToParent)) {
         presentation.setEnabled(false);
-        presentation.setDescription("Can't rebase merged branch '${branchName.get()}'");
+        presentation.setDescription("Can't rebase merged branch '${branch.get().getName()}'");
 
       } else if (branch.get().isNonRootBranch()) {
         var nonRootBranch = branch.get().asNonRootBranch();
         IGitMacheteBranch upstream = nonRootBranch.getUpstreamBranch();
-        presentation.setDescription("Rebase '${branchName}' onto '${upstream.getName()}'");
```
Comment (PawelLipski) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRebaseBranchOntoParentAction.java:94]: Whoops, that was a bug :beetle:

File: config/checker/FORBIDDEN-METHODS.astub [hunk 11]
```diff
@@ -0,0 +1,55 @@
+
+import org.checkerframework.checker.fenum.qual.Fenum;
+
+// Note that in general the primary way to forbid a method/class should be via Checkstyle (`Regexp` -> `illegalPattern`);
+// this is very crude (just detecting a string regardless of its actual semantics) but also very effective and hassle-free for most cases.
```
Comment (mkondratek) [inline config/checker/FORBIDDEN-METHODS.astub:5]: cool :rocket:

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRebaseBranchOntoParentAction.java [hunk 12]
```diff
@@ -106,10 +102,12 @@ public void actionPerformed(AnActionEvent anActionEvent) {
...
     var branchName = getNameOfBranchUnderAction(anActionEvent);
     var branch = branchName.flatMap(bn -> getGitMacheteBranchByName(anActionEvent, bn));
-    if (branch.isDefined() && branch.get().isNonRootBranch()) {
-      doRebase(anActionEvent, branch.get().asNonRootBranch());
+    if (branch.isEmpty()) {
+      LOG.warn("Skipping the action because the branch is undefined");
+    } else if (branch.get().isRootBranch()) {
+      LOG.warn("Skipping the action because the branch '${branch.get().getName()}' is a root branch");
     } else {
-      LOG.warn("Skipping the action because branch is undefined or is a root branch: branch='${branch}'");
```
Comment (mkondratek) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRebaseBranchOntoParentAction.java:112]: yeah, we must be precise :+1:

File: build.gradle (hunk 1 above)
Comment (PawelLipski) [reply build.gradle:133]: Done on next PR

**Pull request merged**